dbms/src/DataStreams/MergeSortingBlockInputStream.cpp

```diff
@@ -304,4 +305,138 @@ void MergeSortingBlockInputStream::remerge()
+
+    bool operator() (size_t a, size_t b) const
+    {
+        for (auto it = left_columns.begin(), jt = right_columns.begin(); it != left_columns.end(); ++it, ++jt)
+        {
+            int res = it->second.direction * it->first->compareAt(a, b, *jt->first, it->second.nulls_direction);
+            if (res < 0)
+                return true;
+            else if (res > 0)
+                return false;
+        }
+        return false;
+    }
+};
+
+Block FinishSortingBlockInputStream::readImpl()
+{
+    if (limit && total_rows_processed == limit)
+        return {};
+
+    Block res;
+    if (impl)
+        res = impl->read();
+
+    /// If res block is empty, we finish sorting previous chunk of blocks.
+    if (!res)
+    {
+        if (end_of_stream)
+            return {};
+
+        blocks.clear();
+        if (tail_block)
+            blocks.push_back(std::move(tail_block));
+
+        Block block;
+        size_t tail_pos = 0;
+        while (true)
+        {
+            block = children.back()->read();
+
+            /// End of input stream, but we can`t return immediatly, we need to merge already read blocks.
+            /// Check it later, when get end of stream from impl.
+            if (!block)
+            {
+                end_of_stream = true;
+                break;
+            }
+
+            // If there were only const columns in sort description, then there is no need to sort.
+            // Return the blocks as is.
+            if (description_to_sort.empty())
+                return block;
+
+            size_t size = block.rows();
+            if (size == 0)
+                continue;
+
+            removeConstantsFromBlock(block);
+
+            /// Find the position of last already read key in current block.
+            if (!blocks.empty())
+            {
+                const Block & last_block = blocks.back();
+                auto last_columns = getColumnsWithSortDescription(last_block, description_sorted);
+                auto current_columns = getColumnsWithSortDescription(block, description_sorted);
+
+                Less less(last_columns, current_columns);
+
+                IColumn::Permutation perm(size);
+                for (size_t i = 0; i < size; ++i)
+                    perm[i] = i;
+
+                auto it = std::upper_bound(perm.begin(), perm.end(), last_block.rows() - 1, less);
+                if (it != perm.end())
+                {
+                    tail_pos = it - perm.begin();
+                    break;
+                }
+            }
+
+            /// If we reach here, that means that current block is first in chunk
+            /// or it all consists of rows with the same key as tail of a previous block.
+            blocks.push_back(block);
+        }
+
+        /// We need to save tail of block, because next block may starts with the same key as in tail
+        /// and we should sort these rows in one chunk.
+        if (block)
```

Comment:
Suggestion: maybe it is better to move this code inside the loop before `break` - this way it is more evident that the block is cut in two only if it contains the next key.